Adding docs.cask.co and supporting files and scripts.

awholegunch opened the pull request:
This provides the scaffolding for the Amazon S3 implementation of documentation hosting. It does not have the data in `www` directory, with the exception of the `resources` directory, as it made the PR too unworkable. The data will be added in a later PR.

Note that most of `builder.py` is not new; it's the same as the previous version in the `documentation` repo, but with a new interface to call it.

Commit:
Adding docs.cask.co and supporting files and scripts.

Commit:
Clarify notes in README.

Commit:
Update script documentation.

Commit:
Remove most of the data files, just leaving the resources.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -0,0 +1,96 @@
+cd $(cd $(dirname ${BASH_SOURCE[0]}); pwd -P)
+
+JSON_VERSIONS_JS="json-versions.js"
+SCRIPT=`basename $0`
```
`$( )` instead of backticks

awholegunch replied:
Fixed.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -0,0 +1,96 @@
+JSON_VERSIONS_JS="json-versions.js"
+SCRIPT=`basename $0`
+SCRIPT_PATH=$(pwd -P)
+TYPES="cdap coopr tigon"
```
Generate this from the configs. `TYPES=$(cd ${SCRIPT_PATH}/../config; ls -1 | sed -e 's/.txt//')`

awholegunch replied:
Fixed.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -0,0 +1,96 @@
+function usage() {
+  echo "Building script for version menu for documentation."
+  echo
+  echo "Usage: ${SCRIPT} <option>"
```
This script only really does one thing... why not just default it to `all` unless a specific type is given? Invocation is:
`${SCRIPT} [type]`

awholegunch replied:
How would you display the usage then? `-h` or `--help`?

awholegunch replied:
So I have changed this to be

```
  echo "Usage: ${SCRIPT} type"
  echo
  echo "Where type is one of:"
  echo
  echo "all"
  echo "${TYPES}"
```

because that's the only way I could see that we could have `build.sh` by itself generate the usage.

wolf31o2 replied:
This build will always only run the "all" case. There just no point in making it conditional. There's no reason to have just the command output the help/usage. Simply make any type that isn't in TYPES be an error/show usage.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -0,0 +1,96 @@
+    echo
+  fi
+  return ${warnings}
+}
```
I would merge these two into a single function, which takes a list of types as arguments. Then, you default the variable you pass to include all types.

```
__list=${1:-${TYPES}}
this_function_name ${__list}
```

Then inside this function, you put your `for` loop, but it's just `for ${type} in ${@}` and loop through all the arguments.

awholegunch replied:
I've still ended up with an `all`, but I did reduce the functions and moved everything into the case statement.

wolf31o2 commented on docs.cask.co/scripts/builder.py:
```diff
@@ -0,0 +1,332 @@
+# JSON Creation (json-versions.js)
+# Creates a JSON file with timeline and formatting information from the data in a 
+# supplied configuration file
+# conf.py ("versions_data")
```
Match the actual config.

awholegunch replied:
Re-written.

wolf31o2 commented on docs.cask.co/scripts/builder.py:
```diff
@@ -0,0 +1,332 @@
+
+    parser = OptionParser(
+        usage="%prog source target",
+        description="Builds a timeline and menu in Javascript format for a web page.")
```
Looks like this is now config / target ?

awholegunch replied:
Fixed.

wolf31o2 commented on docs.cask.co/scripts/builder.py:
```diff
@@ -0,0 +1,332 @@
+    versions_data = configuration["versions_data"]
+    older = versions_data["older"]
+    rev_older = list(older)
+    rev_older.insert(0, versions_data["current"])
```
Use single quotes anywhere where you aren't using string interpolation... `versions_data = configuration['versions_data']`, etc

awholegunch replied:
Fixed. I usually do (now), but I hadn't fixed up this older file everywhere...

Commit:
Changes per comments. Updated README, consolidated Bash script, fixed
Python script error and handling of missing directories. Converted all
simple Python strings to be enclosed in single quotes.

Commit:
Correct usage.

Commit:
Simplification.

Commit:
Improve error handling.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -0,0 +1,86 @@
+  echo
+  usage
+  exit 1
+}
```
So `die` now outputs usage? Why mix these?

awholegunch replied:
Removed.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -0,0 +1,86 @@
+  for type in ${@}; do
+    if [[ -z ${type} ]]; then
+      die "Type needs to be provided."
+    fi
```
Useless code. If `${@}` is empty, the `for` will fail with a syntax error. `for type in ; do`

awholegunch replied:
Moved to where it would do some good.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -0,0 +1,86 @@
+case "${1}" in
+  all )  build_jsons ${TYPES}; exit $?;;
+  *   )  build_jsons ${1}; exit $?;;
+esac
```
This entire script, simplified:

``` bash
#!/usr/bin/env bash
#... header here...

__site=$(cd $(cd $(dirname ${BASH_SOURCE[0]}); cd $(pwd -P)/..; pwd -P) # traverse symlinks, etc
__types=${@:-$(cd ${__site}/configs; ls -1 | sed -e 's/.txt$//')} # support multiple types
for ${__type} in ${__types}; do
  python "${__site}/scripts/builder.py" \
    "${__site}/configs/${__type}.txt" \
    "${__site}/www/${__type}/json-versions.js" || (
      echo "Could not create 'json-versions.js' and 'version' file for '${__type}'" ; exit 1)
done
```

Commit:
Revised script to write to 'target' and have a 'test' option for checking results.

Commit:
Adds copying of any non-current files to the build of the website.
This allows a RM to check in a file and have it be synced automatically.

Commit:
Edits per comments.

wolf31o2 commented on docs.cask.co/scripts/build.sh:
```diff
@@ -19,134 +19,16 @@
+
+rm -rf ${__site}/target || ( echo "Could not remove target directory" ; exit 1)
+mkdir ${__site}/target || ( echo "Could not create target directory" ; exit 1)
+cp -R ${__site}/www/* ${__site}/target || ( echo "Could not copy www to target directory" ; exit 1)
```
Are you wanting to exclude dot files? Otherwise, just `rm -rf ${__site}/target` then `cp -R ${__site}/www ${__site}/target` and skip the mkdir.

awholegunch replied:
Yes, I want to exclude dot files. Especially the Mac `.DS_Store` ones.

wolf31o2 reviewed (approved):
One comment, non-blocking. 👍 LGTM
